Encrypted Logs Polish

oguzkocer opened the pull request:
This PR is a collection of improvements for encrypted logging. There are a bunch of small changes, so it's hard to put them all under a header, but here are some cliff notes:
* Better error handling: I've added new error types and a new `sealed class` for failure types to make it easier to understand what each error means to us. I've also decided to dispatch failed upload actions even if we'll retry them so we can quit early in tests.
* If the log file is missing we now let the client know about it instead of silently deleting it. This ties back to error handling improvements.
* Improved DI for encrypted logging public key. The example apps inject a random key which means the logs will not be available in MC. WPAndroid injects a proper public key.
* Updated `LogEncrypter` to work with `String` instead of `File`. That gives us an easier way to handle file specific issues in the Store. Since we were reading the contents of the file to memory even before the change, I don't think it'll have any impact.
* Added a 30 seconds delay in between uploads. I am still not sure how to go about this and 30 seconds felt pretty long while testing. I am leaving it in the PR for now so we can discuss it.

**Known issues**
1. Even though [we check for the file's existence](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/feature/encrypted-logs-polish/fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt#L127), I sometimes have a crash in a real device for the missing file [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/feature/encrypted-logs-polish/fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt#L140): https://sentry.io/share/issue/5be303a9982544ce852cd63ae2f5293f/. This hasn't happened on an emulator. I don't know how this is even possible with the `exists()` check before reading the contents. I'd have to look into this one, but want to get the other improvements out of the way first.
2. Log files randomly go missing on device if we can't upload them immediately. This feels like an issue with our log file handling in WPAndroid, so I'll look into that.
3. The uploaded log files still show up as a single line in MC. I had a very poor attempt at fixing it in this PR [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/feature/encrypted-logs-polish/fluxc/src/main/java/org/wordpress/android/fluxc/model/encryptedlogging/LogEncrypter.kt#L21) but I haven't yet prioritized looking into it.

jkmassel commented on fluxc/src/main/java/org/wordpress/android/fluxc/model/encryptedlogging/LogEncrypter.kt:
```diff
@@ -2,31 +2,23 @@ package org.wordpress.android.fluxc.model.encryptedlogging
-    }
-
-    fun encrypt(): String = buildString {
+    fun encrypt(text: String): String = buildString {
```
This signature is definitely nice (and I think reads better at the call site), but it introduces an issue with `state`. It can't be re-used between encryptions, so if somehow you get in a situation where you're re-using a `LogEncrypter` across multiple log files, you'll have problems.

One solution might be to just move it inside `encrypt` and pass it around as needed?

oguzkocer replied:
Since we were creating the `LogEncrypter` with a `uuid`, I don't think this would have been an issue. However, that is a fair concern and I admit I didn't really try to improve the `LogEncrypter` very much and it remained in an in-between state with what we had in WPAndroid and what had to change in FluxC. Anyway, a6b16c24b841a6a4ffe9021e288d5b6209af53b1 should address the issue and I am guessing that's what you had in mind.

This also made me think, why not make this a reusable component without any state. So, I followed it up with 8323f8bcd0cd7a43ed3cfb1ff4457590dc038170 which makes `LogEncrypter` reusable and injectable. I am still not super happy with the `encrypt` method signature:
```
    fun encrypt(text: String, uuid: String): String = buildString {
```
Particularly, the `uuid` feels a bit weird there even with the documentation for it. However, I don't have a good idea to resolve that just yet. Let me know if you have any suggestions.

jkmassel replied:
>Anyway, a6b16c2 should address the issue and I am guessing that's what you had in mind.

Yep – this makes it so each log has its own `state`, even if the `LogEncrypter` is reused (which it could be – especially now that it's injectable – we use the same key for the whole app). Mostly I like this because we can't inadvertently break the security by reusing the class incorrectly.

> This also made me think, why not make this a reusable component without any state. So, I followed it up with 8323f8b which makes LogEncrypter reusable and injectable.

This is very nice!

>I am still not super happy with the encrypt method signature

Yeah, I can see how it feels a little odd – I can see why it might be like "why do I care about this UUID thing?". This is why (I think...) at some point I had an intermediate object that represented what was essentially an encryption request – it encapsulated the "we really need a UUID _and_ the text to do this thing" in a way that was a bit more in this specific spot, but had the tradeoff of more overhead.

I can't think of a solution that doesn't involve losing statelessness – we could put the UUID in the constructor and go back to having the `LogEncrypter ` be a single-use object (ie – you throw it away after every use), but I don't think that's nicer. 

The only other thing I was thinking of is that we could rename the method to something like `encryptLogWithLogUUID` (very iOS-ey name...) but that clarifies why the UUID is important?

oguzkocer replied:
Hah - using an intermediate object and renaming the method were 2 of the options I thought about but they didn't feel they were exactly what I was looking for. Let me sleep on it and have another think tomorrow, maybe I'll have a fresh perspective.

jkmassel commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt:
```diff
@@ -31,22 +36,23 @@ import java.io.File
+ * Delay to be used in between uploads whether they are successful or failure. This should help us avoid
+ * `TOO_MANY_REQUESTS` error in typical situations.
+ */
+private const val UPLOAD_DELAY = 30000L
```
This is an interesting approach! I don't mind it, though it keep the user's radio turning on more than we perhaps need?

oguzkocer replied:
@jkmassel Do you mind ellaborating on that a bit? We only use this after a successful or a failed upload, so I don't think we'd be using it enough to cause any concern. We could (should?) check if there are any more logs to be uploaded before calling `uploadNextWithDelay`, but that would only move when the DB check is executed and if there are no more files to upload, we wouldn't proceed in either case. Am I maybe misunderstanding what you meant in this comment?

jkmassel replied:
I think my concern is that we delay for successful requests – IMHO in an optimal world we'd try as many requests as possible as quickly as possible, then back off when we start getting `429` responses. But I can see how the statelessness makes that tricky. Continued in the next comment.

oguzkocer replied:
Oh, we definitely can just try to upload the next one without delay after a successful request. I was only worried about `429`, but as you say, if we get it, we can back off. I'll continue in the next comment as well :)

jkmassel commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt:
```diff
@@ -100,9 +106,8 @@ class EncryptedLogStore @Inject constructor(
-        // TODO: Add a proper back off timing logic
-        delay(10000)
+    private suspend fun uploadNextWithDelay() {
+        delay(UPLOAD_DELAY)
```
This reads far better with a constant, though I wonder about exponential backoff, even with 30s as the starting point?

oguzkocer replied:
Currently `EncryptedLogStore` is stateless with only information coming from either the DB or subsequent requests. Either way, it seems it'd be tricky to carry the delay state across, that's why I opted not to do this at least in the initial version. What do you think, should we still try to get that done? Or maybe is there an easier way I am missing?

jkmassel replied:
Yeah that's tricky – I'm not sure if there's a key-value store or database table you could use to even just have something like `next_upload_attempt_at` where it's like a date or something? Then if you ever hit a `429`, you just add an hour to that? I guess you could even just have that as a column on the encrypted log table, and just add the date to every row? 

I don't think it's critical to have, though it would be nice. My concern is that we may not come back to this functionality for a while (unless we want to tackle the retry stuff as part of integration with Woo and Simplenote?) so I don't want to forget to do it in the future 😅 

Though I would be fine to ship this as-is if we know we'll come back to it :)

oguzkocer replied:
Fair point. I want to make a list of immediate improvements for us to work on after we ship this in this release, so I am not too worried about not coming back to it, but if we can get it resolved now, why not. Let me see if I can find a nice way to do this tomorrow.

Keeping a `dontTryToUploadUntil` date would be very handy since we can use the `429` error to delay 1 hour and not randomly try to upload until that hour is up. The errors were happening much more frequently during development, but now that we have a much better error handling, I don't think we need a fancy backoff timing if we can keep a date like that. We just need to work around 429 and internet connectivity and that'll probably be it.

oguzkocer commented:
Thanks for the review @jkmassel! I've addressed your comments, so this should be ready for another review. Here are the changes since your review:

1. Addressed your comment about state in `LogEncrypter` in a6b16c24b841a6a4ffe9021e288d5b6209af53b1 and 8323f8bcd0cd7a43ed3cfb1ff4457590dc038170. Explained these changes in detail [in this comment](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1650#discussion_r466358283).
2. Replied to your other 2 comments without any changes (yet)
3. Hopefully fixed the missing newline issue in e243fde21569b239782c9f5e920dd03d087ec7b3 thanks to your heads up about `lines()` on Slack. This will hopefully address `Known issue 3`.
4. Merged `develop` into this branch in ccac4e9a55d307f464f06bc6f5fbd50f659ce2a3 as WPAndroid was failing to build without it. I messed up the conflict resolution and had to add a `}` in a20636ed242d2887eda0fb69b9a70114e0447ed6.
5. As per our discussion on Slack, I've added a `canRead` check for the log file in aae84c54812623fa1f820ca02a7304a3f2d86e56. This will hopefully address `Known issue 1`. Also in the same commit, I decided to emit the change to the clients if we fail without queuing the upload to close the loop.

Assuming all this works, the only remaining issue is `Known issue 2` which is happening because we only keep 3 log files at a time as can be seen [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/WordPress.java#L267). This will be fixed in the WPAndroid PR.

oguzkocer commented:
Thanks for the re-review @jkmassel! I addressed your performance comment in https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/e243fde21569b239782c9f5e920dd03d087ec7b3#commitcomment-41258799, if you're good with that, my preference would be to merge this one in and open a new PR for the delay improvements if I can get that done. If I can't, we'll just do it after it's all merged into `develop` so it'll make it to the next code freeze. Does that work for you?

oguzkocer commented:
@jkmassel I am going to assume you're fine with https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1650#commitcomment-41258799 and merge this one. Please let me know if you feel otherwise and I'll address it as soon as possible.
